src/plugins/profiling/server/routes/search_flamechart.ts

```diff
@@ -152,25 +152,47 @@ async function queryFlameGraph(
     }
   );
 
-  const totalCount: number = resEvents.body.aggregations?.total_count.value;
-  const stackTraceEvents = new Map<StackTraceID, number>();
-
+  let totalCount: number = resEvents.body.aggregations?.total_count.value;
   let docCount = 0;
-  let bucketCount = 0;
+
+  const stackTraceEvents = new Map<StackTraceID, number>();
   resEvents.body.aggregations?.group_by.buckets.forEach((item: any) => {
     const traceid: StackTraceID = item.key.traceid;
     stackTraceEvents.set(traceid, item.sum_count.value);
-    docCount += item.doc_count;
-    bucketCount++;
+    docCount++;
   });
-  logger.info('query time registered by ES: ' + resEvents.body.took + 'ms');
   logger.info('events fetched: ' + docCount);
   logger.info('total count: ' + totalCount);
-  logger.info('unique stacktraces: ' + bucketCount);
+  logger.info('unique stacktraces: ' + stackTraceEvents.size);
+
+  // Manual downsampling if totalCount exceeds sampleSize by 10%.
+  if (totalCount > sampleSize * 1.1) {
+    const p = sampleSize / totalCount;
+    totalCount = 0;
+
+    logger.info('downsampling with p=' + p);
+
+    for (const [id, count] of stackTraceEvents) {
+      let newCount = 0;
+      for (let i = 0; i < count; i++) {
+        if (Math.random() < p) {
+          newCount++;
+        }
+      }
+      if (newCount) {
+        stackTraceEvents.set(id, newCount);
+        totalCount += newCount;
+      } else {
+        stackTraceEvents.delete(id);
+      }
+    }
+    logger.info('downsampled total count: ' + totalCount);
+    logger.info('unique downsampled stacktraces: ' + stackTraceEvents.size);
+  }
```

Comment:
should we put this into a dedicated function?
It should be easier to remove when we have random query sampling in ES